### ISSUE TITLE
[inductor] `x + 0` / `x * 1` / `x / 1.0` as a graph output returns the input tensor itself (`out is x`), so an in-place update of the result silently corrupts the input

### ISSUE THREAD

**Issue by cuiliaomei-beep** [pytorch/pytorch#197893]
### 🐛 Describe the bug

**Summary.** For a function whose output is `x / temperature`, `buf * momentum` or `x + bias` called with the neutral value (`1.0`, `1.0`, `0`), eager and `aot_eager` return a fresh tensor, while Inductor returns the input object itself (`out is x`; with `requires_grad`, another object on the same storage), on torch 2.14.0 and nightly, CPU and CUDA. A caller that then updates the result in place (`new_buf.add_(g)`, `logits.clamp_(...)`) overwrites its own input with no error, only for the value that makes the operation a no-op: the same code is correct for `temperature=0.9` and wrong for `temperature=1.0`. Because `joint_graph.remove_no_ops` replaces the node by its operand without the check that `post_grad.remove_noop_ops` has ("Don't introduce new aliasing between inputs and outputs"), every graph that returns such a no-op of an input is affected.

```python
import torch

def scale(logits, temperature):
    return logits / temperature

def momentum_step(buf, momentum):
    return buf * momentum

def add_bias(x, bias):
    return x + bias

for name, f, arg in (("logits / 1.0", scale, 1.0), ("logits / 0.9", scale, 0.9),
                     ("buf * 1.0", momentum_step, 1.0), ("x + 0", add_bias, 0)):
    for backend in ("eager (no compile)", "aot_eager", "inductor"):
        torch._dynamo.reset()
        g = f if backend.startswith("eager") else torch.compile(f, backend=backend)
        x = torch.ones(3)
        out = g(x, arg)
        same = out is x
        out.add_(5)                     # the caller updates the *result* in place
        print(f"{name:13s} {backend:19s} out is x: {str(same):5s}  x afterwards: {x.tolist()}")
```

```
logits / 1.0  eager (no compile)  out is x: False  x afterwards: [1.0, 1.0, 1.0]
logits / 1.0  aot_eager           out is x: False  x afterwards: [1.0, 1.0, 1.0]
logits / 1.0  inductor            out is x: True   x afterwards: [6.0, 6.0, 6.0]
logits / 0.9  eager (no compile)  out is x: False  x afterwards: [1.0, 1.0, 1.0]
logits / 0.9  aot_eager           out is x: False  x afterwards: [1.0, 1.0, 1.0]
logits / 0.9  inductor            out is x: False  x afterwards: [1.0, 1.0, 1.0]
buf * 1.0     eager (no compile)  out is x: False  x afterwards: [1.0, 1.0, 1.0]
buf * 1.0     aot_eager           out is x: False  x afterwards: [1.0, 1.0, 1.0]
buf * 1.0     inductor            out is x: True   x afterwards: [6.0, 6.0, 6.0]
x + 0         eager (no compile)  out is x: False  x afterwards: [1.0, 1.0, 1.0]
x + 0         aot_eager           out is x: False  x afterwards: [1.0, 1.0, 1.0]
x + 0         inductor            out is x: True   x afterwards: [6.0, 6.0, 6.0]
```

Also affected: `0 + x`, `x - 0`, `x + 0.0`, `torch.add(x, 0)`, `(x + 0) + 0`, the same operations with a constant tensor built inside the function (`x * torch.ones(3)`, `x + torch.zeros(3)`, `x - torch.zeros_like(x)`, `x / torch.ones_like(x)`, `x * torch.full_like(x, 1.0)`), integer dtypes, an `nn.Module` whose `forward` is `logits / self.t` with `self.t = 1.0`, and inputs with `requires_grad=True` (there `out is x` is `False` but `out.data_ptr() == x.data_ptr()`, so `out.detach().mul_(2)` doubles `x`). Not affected: `x ** 1`, `x // 1`, `x.clone()`, `x + torch.zeros(())`.

A second, smaller consequence of the same elimination: `x + 0` no longer normalises negative zero (`torch.tensor(-0.0) + 0` is `0.` in eager and `-0.` compiled).

**Root cause and possible fix.** `torch/_inductor/fx_passes/joint_graph.py::remove_no_ops` → `replace_no_op` does `node.replace_all_uses_with(replacement)` whenever shape/dtype/device match and the replacement is not mutated (the guard added for #174187). When `replacement` is a placeholder and `node` feeds the graph output, the compiled graph returns its input, but AOTAutograd has already recorded that output as a non-alias, so the runtime wrapper hands the input tensor back to the caller as if it were fresh. `post_grad.py::remove_noop_ops` avoids the same situation with

```python
# Don't introduce new aliasing between inputs and outputs.
if (not node_is_view and node_storage in output_storages
        and (src_storage in input_storages or src_storage in output_storages)):
    continue
```

Applying the same condition in `replace_no_op` (skip, or replace the node by `aten.clone(replacement)` so the arithmetic still disappears) restores eager semantics; the check also has to cover two outputs that would become the same tensor (`return x * 1, x + 0`).

### Error logs

No error is raised.

### Versions

```
PyTorch version: 2.14.0+cpu
OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
MSVC: Microsoft (R) C/C++ Optimizing Compiler Version 19.44.35228 for x64
Python version: 3.14.7 (64-bit runtime)
Is CUDA available: False
CPU: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
```

Same output on nightly (2.15.0.dev20260911 on Windows, 2.15.0.dev20260920 on Linux), on Linux CPU with 2.14.0, and on CUDA (Tesla T4, 2.14.0+cu130), so the behaviour comes from the shared FX pass and not from a backend.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by chinmayirm** [comment]
I’d like to take this.

`remove_no_ops` in `joint_graph.py` is meant to drop algebraic identities (`x + 0`, `x * 1`, `x / 1`). That is correct while the result is still an interior node: users of the add just start reading `x`, and nothing about the graph’s calling convention changes.

The failure is when that node **is** the graph output. `replace_no_op` does `node.replace_all_uses_with(replacement)` and erases the add/mul/div. If `replacement` is the placeholder for the input, the compiled wrapper returns that same Python object, so `out is x`. Eager always allocates a new tensor for `x + 0`, so `out.add_(...)` is safe there. Under inductor it writes through to the caller’s buffer — and only when the scalar is the identity (`temperature=1.0`), which is why `0.9` looks fine.

This is already treated as a hard invariant. `post_grad.remove_noop_ops` records input and output storages with `get_node_storage` and refuses a rewrite that would make them overlap. `fx_passes/README.md` says the same thing: a pass must not introduce input/output aliasing that was not in the original graph. `joint_graph` never does that check. The existing `is_mutated(replacement)` guard is #174187 (don’t fold away an implicit copy before an in-place); it does not cover aliasing, so both checks need to stay.

Proposed approach
In `remove_no_ops`, collect placeholder storages and output storages the same way `remove_noop_ops` does. In `replace_no_op`, before `replace_all_uses_with`: if the node is not already a view of the replacement, its storage is an output storage, and the replacement’s storage is an input or output storage, leave the node in the graph. Interior `x + 0` should still fold. Do not disable the pass.

Tests: `logits / 1.0`, `buf * 1.0`, `x + 0` — compiled `out is not x`, and `out.add_(5)` does not change `x`. Same ops with `0.9` still fold. With `requires_grad`, `out.data_ptr() != x.data_ptr()`. Also `0 + x`, `x - 0`, and `x * torch.ones_like(x)`.

I would be happy to raise a PR if assigned. Thanks!

**Comment by wwoosshh** [comment]
Scatter variant of the same class of problem, reproduced on 2.14.0+cu130 and 2.15.0.dev20260921+cu130 (Inductor, CUDA; `aot_eager` is correct):

    def fn(x):
        y = torch.diagonal_scatter(x, x.diagonal())   # value-wise a no-op, a fresh tensor in eager
        x.fill_(2)
        return y                                       # inductor returns all 2s

Same for torch.select_scatter(x, x.select(0, 1), 0, 1) and torch.slice_scatter(x, x[1:3], 0, 1, 3).
Here the alias does not come from remove_noop_ops but from the reinplacing pass writing the scatter
result into the input buffer (reinplace_inplaceable_ops_core.can_inplace); the later fill_ then
overwrites it. Also `y = x.neg(); ...; y.copy_(x); return y` makes the returned y alias x.
PR with a fix and a regression test in test/inductor/test_inplacing_pass.py: #198096 (issue #198094)
